File: src/main/java/org/example/expert/domain/user/repository/UserQueryRepositoryImpl.java
package org.example.expert.domain.user.repository;

import com.querydsl.core.types.Projections;
import com.querydsl.jpa.impl.JPAQueryFactory;
import lombok.RequiredArgsConstructor;
import org.example.expert.domain.user.dto.response.UserResponse;
import org.example.expert.domain.user.entity.QUser;

import java.util.List;

@RequiredArgsConstructor
public class UserQueryRepositoryImpl implements UserQueryRepository{

    private final JPAQueryFactory queryFactory;
    private final QUser user = QUser.user;

    @Override
    public List<UserResponse> findUserByNickname(String nickname) {
        return queryFactory
                .select(Projections.constructor(UserResponse.class,
                        user.id,
                        user.email,
                        user.nickname
                ))
                .from(user)
                .where(user.nickname.eq(nickname))
                .fetch();
    }
}

File: src/main/java/org/example/expert/domain/user/repository/UserQueryRepositoryImpl.kt
package org.example.expert.domain.user.repository

import com.querydsl.core.types.Projections
import com.querydsl.jpa.impl.JPAQueryFactory
import org.example.expert.domain.user.dto.response.UserResponse
import org.example.expert.domain.user.entity.QUser
import org.springframework.stereotype.Repository

@Repository
class UserQueryRepositoryImpl(
    private val queryFactory: JPAQueryFactory
) : UserQueryRepository {
    private val user = QUser.user

    override fun findUserByNickname(nickname: String): List<UserResponse> {
        return queryFactory
            .select(
                Projections.constructor(
                    UserResponse::class.java,
                    user.id,
                    user.email,
                    user.nickname
                )
            )
            .from(user)
            .where(user.nickname.eq(nickname))
            .fetch()
    }
}
